File: ruby/sieve/sieve.rb
# @title sieve
# @author cmdellinger
class Sieve
  # @param input_number [Integer] maximum of range to be tested for primes.
  def initialize(input_number)
    @number = input_number
  end
  
  # Finds primes up to the instance number using the Sieve of
  # Eratosthenes algorithm.
  #
  # @return [Array] prime values in range 2 to instance number.
  def primes()
    return [] if @number < 2
    # build Boolean array to use for sieve with buffer to align indices
    sieve_array = Array.new(2, false) + Array.new(@number-1, true)
    # perform Sieve of Eratosthenes eliminations
    (2..Math.sqrt(@number).to_i).each do |i|
        (i**2..@number).step(i) {|j| sieve_array[j] = false} if sieve_array[i] == true
    end
    # return numbers by corresponding index that are still true
    (2..@number).collect {|index| index if sieve_array[index] == true}.compact
  end
end

# BookKeeping Version
module BookKeeping
    # the version of the tests that are being run
    VERSION = 1
end

File: python/sieve/sieve.py
"""
Exercism.io assignment: sieve.py
written by (GitHub):    cmdellinger
Python version:         2.7

Usage:
    sieve.py <number>...

Implementation of the Sieve of Eratosthenes algorithm to find primes up to a
    given number.

Algorithm (from README):
The algorithm consists of repeating the following over and over:
    - take the next available unmarked number in your list (it is prime)
    - mark all the multiples of that number (they are not prime)
Repeat until you have processed each number in your range.

When the algorithm terminates, all the numbers in the list that have not
been marked are prime.
"""

## ------
## Exercism.io solution
## ------

def sieve(limit):
    ''' uses the Sieve of Eratosthenes algorithm to find primes up to limit.
        this version uses optimizations to speed up the algorithm.
    '''
    from math import sqrt
    # build True/False sieve array (Python evaluates non-zero to True)
    sieve_array = [0,0] + range(2, limit+1)
    # perform Sieve of Eratosthenes
    for i in xrange(2, int(sqrt(limit))+1):
        if sieve_array[i]:
            for j in xrange(i**2, limit+1, i):
                sieve_array[j] = 0
    # filter non-zero results
    return filter(lambda x: x, sieve_array)

## ------
## Command-line implementation
## ------

if __name__ == '__main__':
    from docopt import docopt
    # note: everything is stored by parser as string
    for number in map(int, docopt(__doc__)['<number>']):
        print "Maximum: {}\nPrimes: {}\n".format(number, sieve(number))
